app/helpers/state_helper.rb

```diff
@@ -0,0 +1,33 @@
+module StateHelper
+  def state_for_frontend(document)
+    state = compose_state(document.state_history)
+
+    if document.publication_state == "draft"
+      classes = "label label-primary"
+    else
+      classes = "label label-default"
+    end
+
+    content_tag(:span, state, class: classes).html_safe
+  end
+
+  def compose_state(state_history)
+    if state_history.count <= 1 || latest_state(state_history) == latest_non_draft_state(state_history)
+      latest_state(state_history)
+    else
+      "#{latest_non_draft_state(state_history)} with new draft"
+    end
+  end
+
+  def latest_state(state_history)
+    state_history.flatten.last
```

Comment:
As discussed, would be better to use the highest number as hash ordering is not guaranteed.